libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs

```diff
@@ -443,12 +456,52 @@ private object EvalExpression(string exp, ParserRuleContext context = null, stri
             // just don't want to write evaluationTargetStack.Peek() everywhere
             evaluationTargetStack.Peek();
 
-        private (object value, string error) EvalByAdaptiveExpression(string exp, object scope)
+        private (object value, string error) EvalByAdaptiveExpression(string exp, object scope, ParserRuleContext context)
         {
+            var currentTemplate = CurrentTemplate();
+            if (currentTemplate != null)
+            {
+                var source = currentTemplate.SourceRange.Source;
+                if (Path.IsPathRooted(source) && context != null && lgOptions.OnEvent != null)
+                {
+                    var lineOffset = currentTemplate.SourceRange.Range.Start.Line;
+                    var sourceRange = new SourceRange(context, source, lineOffset);
+                    var expressionRef = new ExpressionContext(exp, sourceRange);
+
+                    var expression = currentTemplate.Expressions.FirstOrDefault(u => u.GetId() == expressionRef.GetId());
+                    if (expression != null)
+                    {
+                        lgOptions.OnEvent(expression, new BeginExpressionEvaluationArgs { Source = source, Expression = exp, Context = currentTemplate.SourceRange.ParseTree });
+                    }
+                }
+            }
+
             var parse = this.ExpressionParser.Parse(exp);
             var opt = new Options();
             opt.NullSubstitution = lgOptions.NullSubstitution;
-            return parse.TryEvaluate(scope, opt);
+            var result = parse.TryEvaluate(scope, opt);
+
+            if (currentTemplate != null)
+            {
+                var source = currentTemplate.SourceRange.Source;
+
+                if (Path.IsPathRooted(source) && lgOptions.OnEvent != null)
+                {
+                    string text;
+                    if (string.IsNullOrEmpty(result.error))
+                    {
+                        text = $"Evaluate expression '{exp}' get result: {result.value}";
+                    }
+                    else
+                    {
+                        text = $"Evaluate expression '{exp}' get error: {result.error}";
+                    }
+
+                    lgOptions.OnEvent(MessageKey, new MessageArgs { Source = source, Text = text, Context = currentTemplate.SourceRange.ParseTree });
```

Comment:
what is this MessageKey used for?

Reply:
it is a placeholder, just mark that this event is an message event.